Modules/_functoolsmodule.c

```diff
@@ -277,30 +277,33 @@ partial_new(PyTypeObject *type, PyObject *args, PyObject *kw)
         assert(PyTuple_Check(pto->args));
     }
 
+    /* process keywords */
     if (pto_kw == NULL || PyDict_GET_SIZE(pto_kw) == 0) {
-        if (kw == NULL) {
-            pto->kw = PyDict_New();
-        }
-        else if (Py_REFCNT(kw) == 1) {
-            pto->kw = Py_NewRef(kw);
-        }
-        else {
-            pto->kw = PyDict_Copy(kw);
-        }
+        pto->kw = PyDict_New();
     }
     else {
         pto->kw = PyDict_Copy(pto_kw);
-        if (kw != NULL && pto->kw != NULL) {
-            if (PyDict_Merge(pto->kw, kw, 1) != 0) {
-                Py_DECREF(pto);
-                return NULL;
-            }
-        }
     }
     if (pto->kw == NULL) {
         Py_DECREF(pto);
         return NULL;
     }
+    if (kw != NULL) {
+        PyObject *key, *val;
+        Py_ssize_t pos = 0;
+        while (PyDict_Next(kw, &pos, &key, &val)) {
+            if (val == phold) {
```

Comment:
Technically, `==` is an implementation detail. I would rather use `Py_Is` here:

```suggestion
            if (Py_Is(val, phold)) {
```

Reply:
It was used initially, but switched to `==` later by recommendation of coredev.

Reply:
Oh! Where did that happen?

Reply:
https://github.com/python/cpython/pull/119827

See initial commits and final result. Somewhere along the way.

And it makes sense from a certain POV. Although `==` is implementation detail, it is not going to change. So there is no need to use macros for this in CPython repo. However, someone who does not need to know internals and is writing 3rd party extension, can use `Py_Is`.

At least it is how I see it.

Reply:
Plenty of people learn the internals *from* reading the CPython source, so it would be nice to make it easier on them :)

I'm not going to bother nitpicking this anyway, LGTM.